### PR TITLE
Hec 960 framework for if api

### PR DESCRIPTION
**Pull request by priyankadhiman3006**
ticket- https://jira.tools.tax.service.gov.uk/browse/HEC-961
We still have some queries pending to be answered. Till then  I think we can start with PR and will update the code as the queries are answered.

### CODE REVIEW

File: app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  private def validateInputData(
+    ctutr: String,
+    startDate: String,
+    endDate: String,
+    correlationId: String
+  ): ValidatedNel[ErrorResult, CompanyAccountingPeriod] = {
+
+    val message    = "Submission has not passed validation."
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala:76]: tiny bit of optimisation here, but this `val` and the regexes below it could come out of this function so that they're not being created every time the function is called

File: app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/Environment.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  case object Clone extends Environment { override def value: String = "clone" }
+  case object Live extends Environment { override def value: String = "live" }
+
+  def getEnvironmentFromString(env: String): Option[Environment] = env match {
+    case "ist0"  => Some(Ist0)
+    case "clone" => Some(Clone)
+    case "live"  => Some(Live)
+    case _       => None
+  }
+  implicit val writes: Writes[Environment]                       = Writes { status =>
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/Environment.scala:35]: do we need a `Writes`? We don't send `Environment` as JSON do we?

File: test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  def badJsonResponseFour(
+    code1: String,
+    code2: String,
+    code3: String,
+    code4: String,
+    message1: String,
+    message2: String,
+    message3: String,
+    message4: String
+  ) = Json.parse(s"""
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala:86]: you could generalise this to something like this:
```scala
def badJsonResponse(codeWithMessages: (String, String)*) = {
  val errorObjects =
    codeWithMessages.toList
      .map{ case (code, reason) : """..."""  }
      .mkString(",")
  Json.parse("""{ "failures" : [ errorObjects ] } """)
}
```

File: test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+class IndividualAccountOverviewControllerSpec extends AnyWordSpec with Matchers {
+
+  lazy val mockCC = stubControllerComponents()
+
+  def fakeRequest(environment: String, correlationId: String) = FakeRequest(
+    method = "GET",
+    uri = "/",
+    headers = FakeHeaders(
+      Seq(("Content-type", "application/json"), ("Environment", environment), ("CorrelationId", correlationId))
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala:41]: you can probably get away with not having content type here - there shouldn't be any I don't think cos there is no body in the request

File: test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+import uk.gov.hmrc.hecstubs.controllers.Testdata._
+
+import java.util.UUID
+import scala.concurrent.Future
+
+class IndividualAccountOverviewControllerSpec extends AnyWordSpec with Matchers {
+
+  lazy val mockCC = stubControllerComponents()
+
+  def fakeRequest(environment: String, correlationId: String) = FakeRequest(
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala:37]: doesn't something like this do what you need?
```scala
FakeReques().withHeaders(...)
```
then you don't need to worry about things like supplying the body

File: app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala (hunk 1 above)
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala:76]: sure 👍 . I was thinking about it and then forgot 🤦‍♀️

File: app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/Environment.scala (hunk 2 above)
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/Environment.scala:35]: No, it's a string

File: test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala (hunk 3 above)
Comment (priyankadhiman3006) [reply test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala:86]: I was planning to do something like this
  def badJsonResponse1(codeWithMessages: (InvalidCode, String)*) = {
    val errorObjects = codeWithMessages.toList
      .map { case (code, reason) =>
        ErrorResult(code, s"Submission has not passed validation. $reason")
      }
    Json.toJson(ErrorResponse(errorObjects))
  }
  Will that be ok?
 I feel more comfortable in case class objects :)

File: app/uk/gov/hmrc/hecstubs/util/ValidationUtils.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+object ValidationUtils {
+  def correlationIdValidation(correlationId: String): ValidatedNel[ErrorResult, String] = Try(
+    UUID.fromString(correlationId)
+  ).toOption match {
+    case Some(id) => Valid(id.toString)
+    case None     =>
+      Validated.invalidNel(
+        ErrorResult(InvalidCorrelationId, "Submission has not passed validation. Invalid header CorrelationId.")
+      )
+  }
```
Comment (meghana-viswanath) [inline app/uk/gov/hmrc/hecstubs/util/ValidationUtils.scala:36]: maybe we can directly match on the `Try` result instead of converting to `Option`? Something like -
```
def correlationIdValidation(correlationId: String): ValidatedNel[ErrorResult, String] = Try(
    UUID.fromString(correlationId)
  ) match {
    case Failure(_) => Validated.invalidNel(
      ErrorResult(InvalidCorrelationId, "Submission has not passed validation. Invalid header CorrelationId.")
    )
    case Success(id) =>Valid(id.toString)
  }
```

File: app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        Validated.invalidNel(
+          ErrorResult(InvalidStartDate, s"$message Invalid query parameter startDate.")
+        )
+
+    val endDateValidation: ValidatedNel[ErrorResult, LocalDate] =
+      if (endDate.matches(dateRegex)) Valid(LocalDate.parse(endDate))
+      else
+        Validated.invalidNel(
+          ErrorResult(InvalidEndDate, s"$message Invalid query parameter endDate.")
+        )
```
Comment (meghana-viswanath) [inline app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala:100]: feel free to ignore, possibly could save a few lines of code by turning this into a function and re-using for both from and to dates

File: test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+          }
+
+          "No Environment value is  passed in the header" in {
+            val result: Future[Result] =
+              controller.individualAccountOverview(validUtr, validTaxYear)(
+                fakeRequest("", UUID.randomUUID().toString)
+              )
+            status(result) shouldBe Status.BAD_REQUEST
+
```
Comment (meghana-viswanath) [inline test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala:119]: do we need another test for when the environment value is missing instead of being an empty string?

File: test/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+        )
+        status(result) shouldBe Status.OK
+        contentAsJson(result) mustBe expectedJson
+      }
+
+      "return bad request " when {
+
+        "One invalid parameter" when {
+
+          "ctutr is unvalid" in {
```
Comment (meghana-viswanath) [inline test/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodControllerSpec.scala:141]:
```suggestion
          "ctutr is invalid" in {
```

File: test/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+            val result: Future[Result] =
+              controller.accountingPeriod(inValidCtUtr(0), inValidStartDate(0), validEndDate)(
+                fakeRequest("live", "correlationId")
+              )
+            status(result) shouldBe Status.BAD_REQUEST
+            contentAsJson(result) mustBe expectedJson
+          }
+        }
+
+        "four invlid paramters" when {
```
Comment (meghana-viswanath) [inline test/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodControllerSpec.scala:365]:
```suggestion
        "four invalid paramters" when {
```

File: app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            val responseJson = Json.toJson(data)
+            logger.info(
+              s"Responding to call for Self-Assessment Individual Account Overview for UTR $utr and  taxYear: $taxYear with JSON: ${responseJson.toString()}"
+            )
+            Ok(responseJson)
+
+        }
+      case None    =>
+        logger.info(
+          s"Call failed  for Self-Assessment Individual Account Overview as Environment as Environment in header is invalid"
```
[the remark this answers is not in the document]
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala:71]: We are still waiting  for the error message confirmation from Ian . I will change it for now :)

File: app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    * @return
+    */
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  private def validateIndividualDetails(
+    utr: String,
+    taxYear: String,
+    correlationId: String
+  ): ValidatedNel[ErrorResult, IndividualAccountOverview] = {
+    val utrRegex  = "^[0-9]{10}$"
+    val yearRegex = "^[0-9]{4}$"
```
[the remark this answers is not in the document]
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala]: This has already commented by Andy and has been updated .

File: app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+      case Some(id) => Valid(id.toString)
+      case None     =>
+        Validated.invalidNel(
+          ErrorResult(InvalidCorrelationId, s"$message Invalid header CorrelationId.")
+        )
+
+    }
+
+    (utrValidation, taxYearValidation, correlationIdValidation)
+      .mapN((utr, taxYear, _) => IndividualAccountOverview(utr, taxYear, ReturnFound))
```
[the remark this answers is not in the document]
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala:116]: when you have to validate multiple things and need all errors from the validation, then yes mapN is the way to go

File: app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/InvalidCode.scala [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.hecstubs.models.accountOverviewDetails
+
+import play.api.libs.json.{JsString, Writes}
+
+sealed trait InvalidCode extends Product with Serializable {
```
[the remark this answers is not in the document]
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/InvalidCode.scala:21]: Those code names all started with Invalid, so i thought it make sense to call them InvalidCode, what do you think?

File: app/uk/gov/hmrc/hecstubs/util/ValidationUtils.scala (hunk 6 above)
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/util/ValidationUtils.scala:36]: It's same thing :) anyhow we have to do pattern matching

File: app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala (hunk 7 above)
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala:100]: i thought it myself as well, but then again it's just one line. So i though even if i convert it into function, i have to call that function at all places and again it will be one line . So i dropped the idea. What do you think?

File: test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala (hunk 3 above)
Comment (AndyHWChung) [reply test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala]: that'd work I think - the only danger with that is if we've implemented the writes instance weirdly on `ErrorResult` or `ErrorResponse` we may not catch it with the unit test if we pick up the same format instance in the actual code. At least with explicitly writing out the JSON we get a stronger degree of certainty that the error responses are being written as we expect.

File: test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala (hunk 8 above)
Comment (priyankadhiman3006) [reply test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala:106]: Yes, will add that once Ian confirm the error message to be displayed

File: test/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodControllerSpec.scala (hunk 10 above)
Comment (priyankadhiman3006) [reply test/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodControllerSpec.scala]: I think i have fixed this in latest commit

File: app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/InvalidCode.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (meghana-viswanath) [reply app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/InvalidCode.scala:21]: I think there's one called `ServerError` which doesn't follow the `Invalid` prefix pattern

[the remark this answers is not in the document]
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/InvalidCode.scala:21]: yes, no worries i can change it's name

File: app/uk/gov/hmrc/hecstubs/util/ValidationUtils.scala (hunk 6 above)
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/util/ValidationUtils.scala:36]: I usually prefer Success and failure when i have to use the exception otherwise i opt for optional value. No worries, i can change that

File: app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala (hunk 1 above)
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala]: updated

File: app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/Environment.scala (hunk 2 above)
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/models/accountOverviewDetails/Environment.scala]: updated

File: test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala (hunk 3 above)
Comment (priyankadhiman3006) [reply test/uk/gov/hmrc/hecstubs/controllers/Testdata.scala]: updated

File: test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala (hunk 4 above)
Comment (priyankadhiman3006) [reply test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala]: updated

File: test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala (hunk 5 above)
Comment (priyankadhiman3006) [reply test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala]: updated

File: app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewController.scala]: updated as of now, will update further on confirmation

File: test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala (hunk 8 above)
Comment (priyankadhiman3006) [reply test/uk/gov/hmrc/hecstubs/controllers/IndividualAccountOverviewControllerSpec.scala:108]: added new test case as of now

File: app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+import uk.gov.hmrc.hecstubs.util.{Logging, ValidationUtils}
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import java.time.LocalDate
+import java.util.UUID
+
+@Singleton
+class AccountingPeriodController @Inject() (cc: ControllerComponents) extends BackendController(cc) with Logging {
+
+  val message    = "Submission has not passed validation."
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala:38]: small thing here: `message` seems quite vague as a name - perhaps something like `badRequestMessage` or `validationErrorMessage` is a bit more descriptive. There's a similar thing in `IndividualAccountOverviewController`

Comment (priyankadhiman3006) [reply app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala]: updated

**Review by AndyHWChung — APPROVED** [review]
